### PR TITLE
enh(ruby) Support for character literals (#2950)

### LINKED ISSUE #2950: (Ruby) Character literal notation not supported for some cases
**Describe the issue**
Only some cases of [Character Literals](https://docs.ruby-lang.org/en/3.0.0/doc/syntax/literals_rdoc.html) seem to be supported.
Support is missing for the following groups of cases (in decreasing importance):
- Single slash character (`?/`), breaks following lines as it gets treated as starting a regex
- Escaped backslash character (`?\\`)
- Non-ascii character (`?あ`)
- Unicode using curly-bracket notation (`?\u{1AF9}`)
- Control and Meta characters (`?\C-a`)

Ruby Character Literals - Actual

**Which language seems to have the issue?**
`ruby`

**Are you using `highlight` or `highlightAuto`?**
`highlight`

**Sample Code to Reproduce**
```ruby
c = ?a       #=> "a"
c = ?abc     #=> SyntaxError
c = ?\n      #=> "\n"
c = ?\s      #=> " "
c = ?\\      #=> "\\"
c = ?\u{41}  #=> "A"
c = ?\C-a    #=> "\x01"
c = ?\M-a    #=> "\xE1"
c = ?\M-\C-a #=> "\x81"
c = ?\C-\M-a #=> "\x81", same as above
c = ?あ      #=> "あ"


c = ?/          #=> /
c = ?\123       # octal bit pattern, where nnn is 1-3 octal digits ([0-7])
c = ?\xA1       # hexadecimal bit pattern, where nn is 1-2 hexadecimal digits ([0-9a-fA-F])
c = ?\uAF09     # Unicode character, where nnnn is exactly 4 hexadecimal digits ([0-9a-fA-F])
c = ?\cx        # control character, where x is an ASCII printable character
c = ?\c\M-x     # meta control character, where x is an ASCII printable character
c = ?\c?        # delete, ASCII 7Fh (DEL)
c = ?\C-?       # delete, ASCII 7Fh (DEL)
```

**Expected behavior**
Ruby Character Literals - Expected

**Additional context**
Seen in this StackExchange/CodeGolf answer: https://codegolf.stackexchange.com/a/217367/25026
```ruby
->(n,g=->c,d{(1..n).map{|i|" "*(n-i)+d+" "*2*(n+i-1)+c}},l=g[?/,e=?\\].reverse){[" "*n+?_*n*2,g[e,?/],l[0..-2],l[-1].sub(/ +(?=\/)/,?_*n*2)]}
```
image

### PR DESCRIPTION
**Pull request by il3ven**
Resolves #2950 

### Changes
Improved the regex to detect character literals.

Here is the new regex with a little reasoning for each.
```js
{
  /* To detect, \nnn octal bit pattern, where nnn is 1-3 octal digits ([0-7]) */
  begin: /\B\?(\\\d{1,3})/
},
{
  /* To detect, \xnn hexadecimal bit pattern, where nn is 1-2 hexadecimal digits ([0-9a-fA-F]) */
  begin: /\B\?(\\x[A-Fa-f0-9]{1,2})/
},
{
  /* To detect,
   * \unnnn         Unicode character, where nnnn is exactly 4 hexadecimal digits ([0-9a-fA-F])
   * \u{nnnn ...}   Unicode character(s), where each nnnn is 1-6 hexadecimal digits ([0-9a-fA-F])
   */
  begin: /\B\?(\\u\{?[A-Fa-f0-9]{1,6}\}?)/
},
{
  /* To detect,
   * \cx or \C-x    control character, where x is an ASCII printable character
   * \M-x           meta character, where x is an ASCII printable character
   * \M-\C-x        meta control character, where x is an ASCII printable character
   * \M-\cx         same as above
   * \c\M-x         same as above
   * \C-\M-x        where x is an ASCII printable character
   */
  begin: /\B\?(\\M-\\C-[\x20-\x7e]|\\M-\\c[\x20-\x7e]|\\c\\M-[\x20-\x7e]|\\M-[\x20-\x7e]|\\C-\\M-[\x20-\x7e])/
},
{
  /* To detect, \C-a or \c? */
  begin: /\B\?(\\c[\x20-\x7e]|\\C-[\x20-\x7e])/
},
{
  /* To detect, ?あ */
  begin: /\B\?(\\?\S)/
}
```
I have referred to the following from the [Ruby documentation](https://docs.ruby-lang.org/en/3.0.0/doc/syntax/literals_rdoc.html).

image
image

**Note:**
`c = ?abc     #=> SyntaxError`  highlights as

### Checklist
- [x] Added markup tests, or they don't apply here because...
- [x] Updated the changelog at `CHANGES.md`
- [ ] Added myself to `AUTHORS.txt`, under Contributors

### CODE REVIEW

File: src/languages/ruby.js [hunk 1]
```diff
@@ -104,10 +104,25 @@ export default function(hljs) {
...
+        begin: /\B\?(\\\d{1,3})/
+      },
+      {
+        begin: /\B\?(\\x[A-Fa-f0-9]{1,2})/
+      },
+      {
+        begin: /\B\?(\\u\{?[A-Fa-f0-9]{1,6}\}?)/
+      },
+      {
+        begin: /\B\?(\\M-\\C-[\x20-\x7e]|\\M-\\c[\x20-\x7e]|\\c\\M-[\x20-\x7e]|\\M-[\x20-\x7e]|\\C-\\M-[\x20-\x7e])/
```
Comment (joshgoebel) [inline src/languages/ruby.js:119]: Can the common portion here not be split out to the end without repeating it for every MC variation?

File: src/languages/ruby.js [hunk 2]
```diff
@@ -104,10 +104,25 @@ export default function(hljs) {
...
-        // \B in the beginning suppresses recognition of ?-sequences where ?
-        // is the last character of a preceding identifier, as in: `func?4`
-        begin: /\B\?(\\\d{1,3}|\\x[A-Fa-f0-9]{1,2}|\\u[A-Fa-f0-9]{4}|\\?\S)\b/
+        begin: /\B\?(\\\d{1,3})/
+      },
+      {
+        begin: /\B\?(\\x[A-Fa-f0-9]{1,2})/
+      },
+      {
+        begin: /\B\?(\\u\{?[A-Fa-f0-9]{1,6}\}?)/
```
Comment (joshgoebel) [inline src/languages/ruby.js:116]: Where does the 1-6 come from here?  Should it not be 4?  What other cases are we trying to cover? Your tests only mention "Unicode character, where nnnn is exactly 4 hexadecimal digits "

**Review by joshgoebel — CHANGES_REQUESTED** [review]
Looks pretty good.

File: src/languages/ruby.js (hunk 2 above)
Comment (il3ven) [reply src/languages/ruby.js:116]:
```
\unnnn         Unicode character, where nnnn is exactly 4 hexadecimal digits ([0-9a-fA-F])
\u{nnnn ...}   Unicode character(s), where each nnnn is 1-6 hexadecimal digits ([0-9a-fA-F])
```
This regexp is accounting for these two cases.

File: src/languages/ruby.js (hunk 1 above)
Comment (il3ven) [reply src/languages/ruby.js]: Yes it can. I missed to see that `[\x20-\x7e]` was common. 😅 I have pushed a new commit.

File: src/languages/ruby.js (hunk 2 above)
Comment (joshgoebel) [reply src/languages/ruby.js:116]: Ah, lets please add a hew test cases for this variant since the everything else is covered so well.

File: src/languages/ruby.js [hunk 3]
```diff
@@ -104,10 +104,25 @@ export default function(hljs) {
...
+      },
+      {
+        begin: /\B\?(\\M-\\C-|\\M-\\c|\\c\\M-|\\M-|\\C-\\M-)[\x20-\x7e]/
+      },
+      {
+        begin: /\B\?(\\c[\x20-\x7e]|\\C-[\x20-\x7e])/
```
Comment (joshgoebel) [inline src/languages/ruby.js:122]:
```suggestion
        begin: /\B\?(\\c[\x20-\x7e]|\\C-[\x20-\x7e])/
        begin: /\B\?(\\c|\\C-)[\x20-\x7e]/
```

File: src/languages/ruby.js [hunk 4]
```diff
@@ -104,10 +104,25 @@ export default function(hljs) {
...
+      },
+      {
+        begin: /\B\?(\\?\S)/
```
Comment (joshgoebel) [inline src/languages/ruby.js:125]:
```suggestion
        begin: /\B\?\\?\S/
```
Is the grouping truly needed?

Comment (il3ven) [reply src/languages/ruby.js:125]: No, it is not needed.

File: src/languages/ruby.js (hunk 2 above)
Comment (il3ven) [reply src/languages/ruby.js:116]: Added the required test cases.

**Comment by joshgoebel** [comment]
@il3ven Thanks fo much!

**Review by joshgoebel — APPROVED** [review]
